quickwit-index-config/src/tag_pruning.rs

```diff
@@ -0,0 +1,478 @@
+    Term {
+        is_present: bool,
+        field: String,
+        value: String,
+    },
+}
+
+/// Records terms into a set of tags.
+///
+/// If no terms are available for a given field, it is still
+/// beneficial to call this method with an empty array for values.
+///
+/// In that case, we will record a tag marking that the field has been
+/// processed, allowing for pruning of the split.
+pub fn append_to_tag_set(field: &str, values: &[String], tag_set: &mut BTreeSet<String>) {
+    tag_set.insert(field_tag(field));
+    for value in values {
+        tag_set.insert(term_tag(field, value));
+    }
+}
+
+/// Represents a predicate over the set of tags associated with a given split.
+#[allow(missing_docs)]
+#[derive(Debug, PartialEq, Clone)]
+pub enum TagFilterAst {
+    And(Vec<TagFilterAst>),
+    Or(Vec<TagFilterAst>),
+    Tag {
+        /// If set to false, the predicate tests for the absence of the tag.
+        is_present: bool,
+        tag: String,
+    },
+}
+
+impl Display for TagFilterAst {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let (is_or, children) = match self {
+            TagFilterAst::And(children) => (false, children),
+            TagFilterAst::Or(children) => (true, children),
+            TagFilterAst::Tag { is_present, tag } => {
+                if !is_present {
+                    write!(f, "¬")?;
+                }
+                write!(f, "{}", tag)?;
+                return Ok(());
+            }
+        };
+        if children.is_empty() {
+            return Ok(());
+        }
+        if children.len() == 1 {
+            write!(f, "{}", children[0])?;
+            return Ok(());
+        }
+        if is_or {
+            write!(f, "(")?;
+        }
+        let mut children_it = children.iter();
+        write!(f, "{}", children_it.next().unwrap())?;
+        for child in children_it {
+            if is_or {
+                write!(f, " ∨ {}", child)?;
+            } else {
+                write!(f, " ∧ {}", child)?;
+            }
+        }
+        if is_or {
+            write!(f, ")")?;
+        }
+        Ok(())
+    }
+}
+
+impl TagFilterAst {
+    /// Evaluates the tag filter predicate over a set of tags.
+    pub fn evaluate(&self, tag_set: &BTreeSet<String>) -> bool {
+        match self {
+            TagFilterAst::And(children) => {
+                children.iter().all(|child_ast| child_ast.evaluate(tag_set))
+            }
+            TagFilterAst::Or(children) => {
+                children.iter().any(|child_ast| child_ast.evaluate(tag_set))
+            }
+            TagFilterAst::Tag { is_present, tag } => tag_set.contains(tag) == *is_present,
+        }
+    }
+}
+
+// Takes a tag AST and simplify it in such a way that the resulting tree:
+// - represents a boolean expression that is equivalent to the original ast
+// - it can be equal to true, or to false, but if it is isn't True and False does not appear in the
```

Comment:
I don't understadand. do you mean ?
```suggestion
// - it can be equal to true, or to false, but if it isn't True and False it does not appear in the
```

Reply:
Ah this rustdoc is obsolete. It does not make sense anymore.